File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/db/DbKeyStore.kt
package xyz.speakeasyapp.app.db

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Stable secret that seeds the [SpeakeasyDb] SQLCipher passphrase.
 *
 * # Why this exists
 *
 * The DB passphrase used to be HKDF-derived directly from the Vouchflow
 * **device token**. The device token is an attestation credential: it
 * rotates on biometric/lock-screen reconfiguration (`KEY_INVALIDATED`)
 * and on reinstall. Tying the data-at-rest key to it meant any one of
 * those events silently re-keyed the database — SQLCipher then refused
 * the old file with `file is not a database (code 26)` and the user
 * lost every conversation, with no recovery path. Reported 2026-05-19
 * (@dinnertray: "lost all my conversations", `session_init_failed`).
 *
 * # What this does instead
 *
 * The passphrase is now derived from a **db root secret** that is
 * generated once and never changes for the life of the install. The
 * secret is kept confidential at rest by wrapping it with an
 * AES-256-GCM key held in the `AndroidKeyStore`:
 *
 *   - The keystore key is hardware-backed where available and is
 *     **not** auth-bound (`setUserAuthenticationRequired` is never
 *     set), so biometric re-enrollment does NOT invalidate it. That is
 *     the whole point — a fingerprint change must not cost the user
 *     their message history.
 *   - The wrapped secret (IV ‖ ciphertext, Base64) lives in a plain
 *     `SharedPreferences` file. It is useless without the keystore key.
 *
 * The secret is app-scoped: it does not survive uninstall or
 * "clear data", which is the intended policy (reinstall = fresh start,
 * same as Signal/WhatsApp — see `backup_rules.xml`).
 *
 * # Migration
 *
 * On the first open after upgrading from the token-derived scheme,
 * [SpeakeasyDb] seeds this store with the *current* device token, so
 * the derivation output is byte-identical to the old key and the
 * existing database opens unchanged. From that point on the secret is
 * frozen and token rotation is irrelevant. If the token is already
 * gone at upgrade time, the legacy DB is unrecoverable anyway and the
 * recovery path in [SpeakeasyDb.open] recreates it.
 */
object DbKeyStore {
  private const val TAG = "DbKeyStore"
  private const val PREFS_FILE = "speakeasy_db_root"
  private const val PREF_KEY = "wrapped_secret"
  private const val KEYSTORE_ALIAS = "speakeasy_db_root_v1"
  private const val ANDROID_KEYSTORE = "AndroidKeyStore"
  private const val AES_GCM = "AES/GCM/NoPadding"
  private const val GCM_TAG_BITS = 128
  private const val GCM_IV_BYTES = 12

  /**
   * Outcome of a root-secret read. [Absent] (no prefs entry) is the safe
   * "fresh install" path where seeding is correct; [Unavailable] (the wrapped
   * secret IS present but unwrap failed — Keystore key invalidated, AEAD tag
   * mismatch) means the secret may still be recoverable and the caller MUST
   * NOT wipe the DB. Conflating these was the cause of "messages deleted on
   * update".
   */
  sealed class LoadResult {
    data class Found(val secret: String) : LoadResult()
    object Absent : LoadResult()
    data class Unavailable(val error: Throwable) : LoadResult()
  }

  /** Read the db root secret, distinguishing "no entry" from "unwrap failed". */
  fun loadResult(context: Context): LoadResult {
    val wrapped =
        context
            .getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
            .getString(PREF_KEY, null) ?: return LoadResult.Absent
    return try {
      LoadResult.Found(unwrap(Base64.decode(wrapped, Base64.NO_WRAP)))
    } catch (e: Throwable) {
      // The wrapped secret IS present but couldn't be decrypted (Keystore key
      // invalidated / AEAD failure / corrupt blob). This is NOT a fresh
      // install — never wipe on it.
      Log.w(TAG, "unwrap failed — secret present but unreadable; will NOT wipe", e)
      LoadResult.Unavailable(e)
    }
  }

  /**
   * The persisted db root secret, or `null` if absent **or unreadable**.
   * Prefer [loadResult] on the open path so a transient unwrap failure isn't
   * mistaken for "fresh install" and used to justify a wipe.
   */
  fun load(context: Context): String? =
      (loadResult(context) as? LoadResult.Found)?.secret

  /**
   * Persist [secret] as the db root secret. Returns `true` iff it is now
   * durably stored (verified by read-back); `false` means the caller must not
   * create a DB it can't re-key next launch.
   */
  fun store(context: Context, secret: String): Boolean {
    return try {
      val wrapped = Base64.encodeToString(wrap(secret), Base64.NO_WRAP)
      val committed =
          context
              .getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
              .edit()
              .putString(PREF_KEY, wrapped)
              .commit()
      committed && load(context) == secret
    } catch (e: Throwable) {
      Log.w(TAG, "store failed — db root secret NOT persisted", e)
      false
    }
  }

  /**
   * Drop the wrapped secret. Used by account deletion alongside
   * [SpeakeasyDb.wipe] so a later re-enrollment starts from a genuinely
   * empty store. The keystore key is left in place — harmless without
   * the ciphertext, and reused if the user re-onboards.
   */
  fun clear(context: Context) {
    context
        .getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
        .edit()
        .remove(PREF_KEY)
        .commit()
  }

  /** GCM-encrypt [secret]; output is IV ‖ ciphertext. */
  private fun wrap(secret: String): ByteArray {
    val cipher = Cipher.getInstance(AES_GCM)
    cipher.init(Cipher.ENCRYPT_MODE, keystoreKey())
    val iv = cipher.iv
    val ct = cipher.doFinal(secret.toByteArray(Charsets.UTF_8))
    return iv + ct
  }

  /** Reverse [wrap]: split IV ‖ ciphertext and GCM-decrypt. */
  private fun unwrap(blob: ByteArray): String {
    val iv = blob.copyOfRange(0, GCM_IV_BYTES)
    val ct = blob.copyOfRange(GCM_IV_BYTES, blob.size)
    val cipher = Cipher.getInstance(AES_GCM)
    cipher.init(Cipher.DECRYPT_MODE, keystoreKey(), GCMParameterSpec(GCM_TAG_BITS, iv))
    return String(cipher.doFinal(ct), Charsets.UTF_8)
  }

  /** Fetch the wrapping key from the keystore, generating it on first use. */
  @Synchronized
  private fun keystoreKey(): SecretKey {
    val ks = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }
    (ks.getEntry(KEYSTORE_ALIAS, null) as? KeyStore.SecretKeyEntry)?.let {
      return it.secretKey
    }
    val generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE)
    generator.init(
        KeyGenParameterSpec.Builder(
                KEYSTORE_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
            .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
            .setKeySize(256)
            // Deliberately NOT setUserAuthenticationRequired(true): the
            // key must survive biometric re-enrollment. See class doc.
            .build())
    return generator.generateKey()
  }
}

File: apps/mobile/ios/SpeakeasyBridges/Db/DbKeyStore.swift
//
//  DbKeyStore.swift
//  Speakeasy
//
//  Stable secret that seeds the SpeakeasyDb SQLCipher passphrase.
//  Mirrors apps/mobile/android/.../db/DbKeyStore.kt.
//
//  # Why this exists
//
//  The DB passphrase used to be HKDF-derived directly from the Vouchflow
//  device token. That token is an attestation credential — it rotates on
//  biometric reconfiguration and reinstall. Tying the data-at-rest key to
//  it meant any such event silently re-keyed the database, SQLCipher then
//  rejected the old file ("file is not a database"), and the user lost
//  every conversation with no recovery path.
//
//  # What this does instead
//
//  The passphrase is derived from a db root secret generated once and
//  frozen for the life of the install. The secret lives in the Keychain
//  as a generic-password item:
//
//    - kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly: readable after
//      the first unlock following a reboot, device-bound, never synced to
//      iCloud. It is NOT biometric-bound, so Face/Touch ID re-enrollment
//      does not invalidate it — the whole point of the decoupling.
//    - The item is app-scoped: it does not survive uninstall, which is
//      the intended policy (reinstall = fresh start).
//

import Foundation
import Security

enum DbKeyStore {
    private static let service = "xyz.speakeasyapp.db"
    private static let account = "root-secret-v1"

    /// Outcome of a root-secret read. The distinction is load-bearing: a
    /// missing item (`absent`) is the safe "fresh install" path where seeding
    /// a new secret is correct, but a *failed read* (`unavailable`) means the
    /// secret may well still exist — re-keying then would WIPE all Signal
    /// sessions + message history for what is often a transient error (device
    /// locked right after a reboot/OS-update → errSecInteractionNotAllowed
    /// -25308; missing keychain entitlement → -34018). The caller MUST NOT
    /// wipe on `unavailable`. Collapsing these two into "nil" was the cause of
    /// "all my messages disappeared after an update" + "couldn't decrypt this
    /// message" (the session the secret protected got wiped).
    enum LoadResult {
        case found(String)
        case absent
        case unavailable(OSStatus)
    }

    private static func baseQuery() -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
    }

    /// Read the db root secret, distinguishing "no item" from "read failed".
    static func loadResult() -> LoadResult {
        var query = baseQuery()
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        switch status {
        case errSecSuccess:
            guard let data = item as? Data, let s = String(data: data, encoding: .utf8) else {
                // Item present but unreadable/garbled — treat as a read
                // failure, NOT as absent. Never wipe on this.
                NSLog("[DbKeyStore] item present but undecodable — treating as unavailable (will NOT wipe)")
                return .unavailable(errSecDecode)
            }
            return .found(s)
        case errSecItemNotFound:
            return .absent
        default:
            NSLog("[DbKeyStore] load failed: OSStatus \(status) — secret may exist; will NOT wipe, open will retry. (-25308 = device locked after reboot/OS update; -34018 = keychain entitlement missing on unsigned simulator builds)")
            return .unavailable(status)
        }
    }

    /// The persisted db root secret, or `nil` if absent **or unreadable**.
    /// Prefer `loadResult()` on the open path so a transient read failure
    /// isn't mistaken for "fresh install" and used to justify a wipe.
    static func load() -> String? {
        if case .found(let s) = loadResult() { return s }
        return nil
    }

    /// Persist `secret` as the db root secret. Overwrites any prior value.
    /// Returns `true` iff the secret is now durably stored (verified by a
    /// read-back); a `false` return means the next launch will re-key and
    /// wipe — the caller logs the loss path.
    @discardableResult
    static func store(_ secret: String) -> Bool {
        SecItemDelete(baseQuery() as CFDictionary)
        var add = baseQuery()
        add[kSecValueData as String] = Data(secret.utf8)
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(add as CFDictionary, nil)
        if status != errSecSuccess {
            NSLog("[DbKeyStore] store failed: OSStatus \(status) — db root secret NOT persisted; history will reset every launch until this succeeds. (-34018 = keychain entitlement missing, common on unsigned simulator builds)")
            return false
        }
        // Read-back guard: confirm the write is actually retrievable. Catches
        // platforms/configs where SecItemAdd reports success but the item is
        // not readable on the next access path.
        if load() == nil {
            NSLog("[DbKeyStore] store succeeded but read-back returned nil — keychain not persisting; history will reset every launch")
            return false
        }
        return true
    }

    /// Drop the stored secret. Used by account deletion alongside
    /// `SpeakeasyDb.wipe()` so a re-enrollment starts from an empty store.
    static func clear() {
        SecItemDelete(baseQuery() as CFDictionary)
    }
}
